### PR TITLE
CircleCI and test optimizations

### PR DESCRIPTION
**Pull request by costinm**
Another round of optimizations for the CI and tests. 

The goal remains to have a reliable and fast set of pre-submits on CI, and run the large tests
on real clusters ( via Prow or vendor-specific code ).

Bookinfo-based tests are heavy and not appropriate for CI and local testing - they run best on a real
cluster with >3 nodes. They kind-of run in CI and minikube - but is not the right environment.

'Reliable' means that if the CI pre-submit is red, you can be confident something is wrong with your PR. 
We separate the tests into 'flaky' and 'stable' - alpha1 egress tests depend on external servers and fail
with no change in the code, so they're moving to the flaky suite and excluded from presubmit.

For greater visibility (in the report) - I'm passing t to the tests, so each sub-test can be run as a separate 
t.Run ( in progress ). A subset of the Pilot tests are now separated - so it is possible to see what exactly 
is failing ( for now: traffic split with regex, in the 'auth' env).

To identify MTLS-related problems, we run the tests with enabled/disabled. The plan is to require
both - but after we confirm that MTLS is no longer flaky.

### CODE REVIEW

File: tests/e2e/tests/pilot/egress_rules_test.go [hunk 1]
```diff
@@ -23,6 +23,8 @@ import (
 
 	multierror "github.com/hashicorp/go-multierror"
 
+	"testing"
```
Comment (ZackButcher) [inline tests/e2e/tests/pilot/egress_rules_test.go]: nit: bad import block, run goimports please.

File: tests/e2e/tests/pilot/grpc_test.go [hunk 2]
```diff
@@ -17,6 +17,8 @@ package pilot
 import (
 	"fmt"
 
+	"testing"
```
Comment (ZackButcher) [inline tests/e2e/tests/pilot/grpc_test.go]: same

File: tests/istio.mk [hunk 3]
```diff
@@ -161,3 +161,22 @@ test/minikube/auth/e2e_pilot_alpha1: istioctl
...
+		-hub ${HUB} -tag ${TAG} \
+		--skip-cleanup --mixer=true \
+		-errorlogsdir=${OUT_DIR}/logs \
+		--use-sidecar-injector=false \
+		--auth_enable=false \
+		-v1alpha3=false -v1alpha1=true \
+		--core-files-dir=${OUT_DIR}/logs \
+        --ns pilot-auth-system \
+        -n pilot-test \
+           ${TESTOPTS} | tee ${OUT_DIR}/tests/test-report-noauth-pilot-v1.raw
```
Comment (ZackButcher) [inline tests/istio.mk]: fix the `tee` here please: `${TESTOPTS} > tee ...`. We just had another PR fixing all of these, I thought.

**Comment by ZackButcher** [comment]
Vadim owns all of egress, him approving is totally appropriate.

**Comment by kyessenov** [comment]
Please don't abuse `t.Run` + `t.Parallel` for this purpose. The test tests are not parallelizable, they need to be executed in the exact sequence with no suspension.

File: tests/istio.mk (hunk 3 above)
Comment (costinm) [reply tests/istio.mk]: The ">" doesn't work well - must have tee (circleci times out if no output in 10 min). 
The trick is the pipefail.

**Comment by costinm** [comment]
make format. 
I noticed that 'rerun' on circleci test fails - due to the 'optimization' in dependencies. However it seems more time is no spend in the optimization than the actual checkout - so moved the circleci test job back to checking out. Will see how it works with test job, move the others over if all is faster.

**Comment by ldemailly** [comment]
I dismissed the review because I want PRs touching the set of tests to be reviewed by at least oncall + owners - we need more, more stable tests, in order to release 0.7, not fewer

File: tests/istio.mk (hunk 3 above)
Comment (ZackButcher) [reply tests/istio.mk]: Ah, thanks for clarifying.

File: tests/e2e/tests/pilot/egress_rules_test.go (hunk 1 above)
Comment (ldemailly) [reply tests/e2e/tests/pilot/egress_rules_test.go]: are the linters broken/not finding those ?

**Comment by ldemailly** [comment]
> I noticed that 'rerun' on circleci test fails - due to the 'optimization' in dependencies. 

the 'optimization' predates the vendor cleanup; cool if you clean it up
(just make sure the submodule is present)

File: .circleci/config.yml [hunk 4]
```diff
@@ -243,8 +244,7 @@ jobs:
       - store_test_results:
           path: /go/out/tests
 
-
-  e2e-pilot-auth:
+  e2e-pilot-alpha3:
```
Comment (ZackButcher) [inline .circleci/config.yml]: `e2e-pilot-alpha3-noauth` instead?

File: .circleci/config.yml [hunk 5]
```diff
@@ -342,7 +342,7 @@ jobs:
             kubectl cluster-info dump > /go/out/logs/cluster-info.dump.txt
             kubectl describe pods -n pilot-auth > /go/out/logs/pods-system.txt
             kubectl describe pods -n pilot-auth-system > /go/out/logs/pods-test.txt
-            /go/bin/go-junit-report </go/out/tests/test-report-auth-pilot-v1.raw > /go/out/tests/test-report-auth-pilot-v1.xml
+            cat /go/out/tests/*.raw | /go/bin/go-junit-report > /go/out/tests/test-report-noauth-pilot.xml
```
Comment (ZackButcher) [inline .circleci/config.yml]: isn't this running auth tests, `make test/minikube/auth/e2e_pilot_alpha1 HUB="${HUB}" TAG="${TAG}"` above?

File: .circleci/config.yml (hunk 4 above)
Comment (ZackButcher) [reply .circleci/config.yml]: Since it executes `make test/minikube/noauth/e2e_pilot HUB="${HUB}" TAG="${TAG}"`

Comment (costinm) [reply .circleci/config.yml]: Will rename them in separate PR.

File: .circleci/config.yml (hunk 5 above)
Comment (costinm) [reply .circleci/config.yml]: Separate PR (given the long wait for a PR to submit)

**Pull request merged**